### PR TITLE
Adding XML Version

### PR DESCRIPTION
**Pull request by aodhan-domhnaill**
# Description

https://github.com/medic/cht-conf/issues/395

medic/cht-core#7561

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [x] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: webapp/src/ts/services/enketo.service.ts [hunk 1]
```diff
@@ -570,6 +571,7 @@ export class EnketoService {
       .map((idx, element) => {
         const docToStore:any = this.enketoTranslationService.reportRecordToJs(getOuterHTML(element));
         docToStore._id = getId(Xpath.getElementXPath(element));
+        docToStore.xmlVersion = xmlVersion;
```
Comment (garethbowen) [inline webapp/src/ts/services/enketo.service.ts:574]: I know this is a draft, but I thought I'd throw a comment in...

This is setting the `xmlVersion` in the child docs being created by the report but not in the report itself. This is potentially useful but we don't store anything about how the child doc was created, not even the form ID, so it's potentially not useful on its own. The primary place this should be set is in the `create` function where the report itself is created, right beside where we set the `form` property.

For CouchDB properties we [use snake case](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/#use-snake_case-for-couchdb-document-property-names) for naming (like `reported_date` below). This applies to both the form and report docs. There are some couterexamples you can find but we're standardising at least on _new_ properties.

Finally I'm not sure `xmlVersion` will make much sense to someone looking at the report who doesn't know the context. I think `form_version` would be better and be obvious when stored beside `form` (the form's ID).

Comment (aidan-plenert-macdonald) [reply webapp/src/ts/services/enketo.service.ts:574]: Thanks. So I understand, the change I made in https://github.com/medic/cht-conf/pull/431 will cause the `get(internalId)` function [here](https://github.com/medic/cht-core/blob/5ae422eb51b2e9bdf1fac49def562387a193348a/webapp/src/ts/services/xml-forms.service.ts#L237) to return a doc Promise that contains the new field `xmlVersion` that I add in https://github.com/medic/cht-conf/pull/431

Is that correct?

Comment (garethbowen) [reply webapp/src/ts/services/enketo.service.ts:574]: Yes it should do! You'll obviously need to get the CHT running in a developer environment to test this end to end.

**Comment by aidan-plenert-macdonald** [comment]
Caught the breaking change. Running with the new version of CHT Conf.

```
 FAILED TESTS:
  Enketo service
    save
      ✖ should pass docs to transitions and save results
        Chrome Headless 93.0.4577.63 (Linux x86_64)
      AssertionError: expected [ Array(4) ] to deeply equal [ Array(4) ]

      + expected - actual

             "lmp": "10"
             "name": "Sally"
           }
           "form": "V"
      -    "form_version": [undefined]
           "from": "555"
           "geolocation": {
             "geo": "data"
           }
```

**Comment by aodhan-domhnaill** [comment]
Unblocked https://github.com/medic/cht-core/blob/master/package.json#L61

**Comment by aodhan-domhnaill** [comment]
```
            "form_version": {
      -      "__zone_symbol__state": true
      -      "__zone_symbol__value": [undefined]
      +      "sha256": "imahash"
      +      "time": "1"
           }
```

**Comment by aodhan-domhnaill** [comment]
```
       -    "form_version": [undefined]
      +    "form_version": {
      +      "sha256": "imahash"
      +      "time": "1"
      +    }
```

File: webapp/src/ts/services/enketo.service.ts [hunk 2]
```diff
@@ -627,6 +627,12 @@ export class EnketoService {
       .then(formDoc => this.getFormAttachment(formDoc));
   }
 
+  private getFormVersion(form) {
+    return this.xmlFormsService
+      .get(form)
+      .then(formDoc => formDoc.xmlVersion);
```
[the remark this answers is not in the document]
Comment (aodhan-domhnaill) [reply webapp/src/ts/services/enketo.service.ts:633]: Okay. I'll take a look at doing that.

File: webapp/tests/karma/ts/services/enketo.service.spec.ts [hunk 3]
```diff
@@ -1520,8 +1520,11 @@ describe('Enketo service', () => {
           </data>`;
       form.getDataStr.returns(content);
 
-      dbBulkDocs.callsFake(docs => Promise.resolve(docs.map(doc => ({ ok: true, id: doc._id, rev: '1' }))));
+      dbBulkDocs.callsFake(docs => Promise.resolve(docs.map(doc => ({
+        ok: true, id: doc._id, form_version: { time: '1', sha256: 'imahash' }, rev: '1'
+      }))));
```
Comment (aodhan-domhnaill) [inline webapp/tests/karma/ts/services/enketo.service.spec.ts:1525]: Assert that database is only called once

**Comment by aodhan-domhnaill** [comment]
```

FAILED TESTS:
  Enketo service
    save
      ✖ should pass docs to transitions and save results
        Chrome Headless 98.0.4758.80 (Linux x86_64)
      AssertionError: expected 0 to equal 1

      + expected - actual

      -0
      +1
```

Looks like dbGet isn't getting called.

**Comment by garethbowen** [comment]
> Looks like dbGet isn't getting called.

In the enketo.service.spec.ts we mock out the `XmlFormsService.get` because we just want to test the Enketo service and not the XmlForms service too. I think what's happening is because this is mocked out the `db.get` is never called regardless. The right unit test would make sure that `XmlFormsService.get` is only called on `render` and not on `save` so we can be sure the final xml version comes from when the form is first shown on the screen.

**Review by m5r — APPROVED** [review]

**Pull request merged**